libs/community/langchain_community/vectorstores/lantern.py

```diff
@@ -0,0 +1,851 @@
+                session.execute(statement)
+                session.commit()
+        except Exception as e:
+            self.logger.exception(e)
+
+    def create_tables_if_not_exists(self) -> None:
+        try:
+            self.create_collection()
+        except ProgrammingError:
+            pass
+
+    def drop_table(self) -> None:
+        try:
+            self.EmbeddingStore.__table__.drop(self._conn.engine)
+        except ProgrammingError:
+            pass
+
+    def drop_tables(self) -> None:
+        self.drop_table()
+
+    def _hamming_relevance_score_fn(self, distance: float) -> float:
+        return distance
+
+    def _select_relevance_score_fn(self) -> Callable[[float], float]:
+        """
+        The 'correct' relevance function
+        may differ depending on a few things, including:
+        - the distance / similarity metric used by the VectorStore
+        - the scale of your embeddings (OpenAI's are unit normed. Many others are not!)
+        - embedding dimensionality
+        - etc.
+        """
+        if self.override_relevance_score_fn is not None:
+            return self.override_relevance_score_fn
+
+        # Default strategy is to rely on distance strategy provided
+        # in vectorstore constructor
+        if self.distance_strategy == DistanceStrategy.COSINE:
+            return self._cosine_relevance_score_fn
+        elif self.distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return self._euclidean_relevance_score_fn
+        elif self.distance_strategy == DistanceStrategy.HAMMING:
+            return self._hamming_relevance_score_fn
+        else:
+            raise ValueError(
+                "No supported normalization function"
+                f" for distance_strategy of {self._distance_strategy}."
+                "Consider providing relevance_score_fn to Lantern constructor."
+            )
+
+    def _get_op_class(self) -> str:
+        if self.distance_strategy == DistanceStrategy.COSINE:
+            return "dist_cos_ops"
+        elif self.distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return "dist_l2sq_ops"
+        elif self.distance_strategy == DistanceStrategy.HAMMING:
+            return "dist_hamming_ops"
+        else:
+            raise ValueError(
+                "No supported operator class"
+                f" for distance_strategy of {self._distance_strategy}."
+            )
+
+    def _get_operator(self) -> str:
+        if self.distance_strategy == DistanceStrategy.COSINE:
+            return "<=>"
+        elif self.distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return "<->"
+        elif self.distance_strategy == DistanceStrategy.HAMMING:
+            return "<+>"
+        else:
+            raise ValueError(
+                "No supported operator"
+                f" for distance_strategy of {self._distance_strategy}."
+            )
+
+    def _typed_arg_for_distance(
+        self, embedding: List[float | int]
```

Comment:
The "|" syntax for typehints isn't supported for python 3.8. Better to use `Union[float, int]`